plugins/woocommerce/src/Internal/Caches/EntityVersionsCache.php

```diff
@@ -0,0 +1,215 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Automattic\WooCommerce\Internal\Caches;
+
+use Automattic\WooCommerce\Proxies\LegacyProxy;
+
+/**
+ * Entity versions cache class.
+ *
+ * Provides a generic mechanism for caching versions of mutable entities.
+ */
+class EntityVersionsCache {
+
+	/**
+	 * Is the entity versions cache enabled?
+	 *
+	 * @var bool|null
+	 */
+	private ?bool $is_enabled = null;
+
+
+	/**
+	 * Tells whether the entity versions cache is enabled or not.
+	 * The other methods of this class should be used only if this method returns true.
+	 *
+	 * @return bool
+	 */
+	public function is_enabled(): bool {
+		if ( ! is_null( $this->is_enabled ) ) {
+			return $this->is_enabled;
+		}
+
+		/**
+		 * Filter whether to enable the entity versions cache.
+		 *
+		 * By default, the entity versions cache will be enabled only if an external
+		 * object cache is configured in WordPress. Enabling it otherwise is not recommended
+		 * (outside of development and testing scenarios) since cached entries would be stored
+		 * directly in the options table, causing high stress in the database.
+		 *
+		 * @since 10.4.0
+		 *
+		 * @param bool   $use_output_cache Whether to use output cache. Default is the result of wp_using_ext_object_cache().
+		 * @return bool True to enable output cache, false to disable.
+		 */
+		$this->is_enabled = apply_filters(
+			'woocommerce_enable_entity_versions_cache',
+			wc_get_container()->get( LegacyProxy::class )->call_function( 'wp_using_ext_object_cache' ) ?? false
+		);
+
+		return $this->is_enabled;
+	}
+
+	/**
+	 * Get the current version of an entity.
+	 *
+	 * @param string     $entity_type Entity type.
+	 * @param string|int $entity_id Entity ID.
+	 * @return string Entity version.
+	 * @throws \InvalidArgumentException If entity_type or entity_id are invalid.
+	 */
+	public function get_entity_version( string $entity_type, $entity_id ): string {
+		$this->validate_input( $entity_type, $entity_id );
+
+		$transient_name = "wc_entity_version_{$entity_type}_{$entity_id}";
+		$version        = $this->get_cached( $transient_name );
+		if ( false === $version ) {
+			$version = $this->modify_entity_version( $entity_type, $entity_id );
+		} else {
+			// Refresh the transient lifetime.
+			$this->store_entity_version( $entity_type, $entity_id, $version, false );
+		}
+		return $version;
+	}
+
+	/**
+	 * Generate and store a new version for an entity.
+	 *
+	 * @param string     $entity_type Entity type.
+	 * @param string|int $entity_id   Entity ID.
+	 * @return string The new entity version.
+	 * @throws \InvalidArgumentException If entity_type or entity_id are invalid.
+	 */
+	public function modify_entity_version( string $entity_type, $entity_id ): string {
```

Comment:
Nitpicky, but it took me a second to grok what/how this function was modifying.  Maybe `generate_new_entity_version()` to match how it's described?